mlir-sycl/doc/design/SYCLKernelSubmissionRepresentation.md

```diff
@@ -0,0 +1,510 @@
+                  _KERNELFUNCPARAM(KernelFunc) __SYCL_ANNOTATE(kernel)) { ... }
+```
+
+In the IR, the lambda object is an `!llvm.struct` whose elements correspond to
+the types of the captured variables. We then look for assignments to the
+top-level elements in the (potentially nested) struct. As this code is generated
+by the compiler frontend, it is idiomatic enough to match for stores and
+`memcpy`s to addresses computed by GEP instructions into the lambda object.
+
+#### Heuristic adaption of values
+
+Unfortunately, there is not always a 1:1 correspondence between assignments and
+captured values. To that end, we apply a set of heuristics to determine the
+actual captured object.
+
+- **Accessors:** The frontend generates a series of stores and `memcpy`s when
+  assigning an accessors (or local accessor), causing us to match only the first
+  member. In order to find a pointer to the entire object, we look through the
+  load of the first member, which should point to an allocation of an
+  `!llvm.struct` representing the `sycl::accessor` class. If we find the
+  corresponding `sycl.host.constructor` op as well, we capture the allocation
+  instead and forward the constructor op's type attribute.
+
+- **Coalesced scalars:** We have observed that the frontend sometimes coalesces
+  sequences of scalar arguments of the same type into vectors, to assign
+  multiple values at once. As this leads to mismatches between captured and
+  expected kernel arguments, we reverse this transformation and instead mark the
+  individual, extracted elements of the vector as captured.
+
+- **Split arrays:** Similarly, the frontend sometimes chooses to pass arrays as
+  a sequence of vectors and scalars, e.g. a 9-element arrays as two 4-element
+  vectors and a scalar. From a set of candidate assignments, we define either a
+  new vector-typed global if all elements are constant, or else reconstruct a
+  vector value on the fly that comprises all previously captured parts.
+
+### Assignment of kernel to handler
+
+We are raising the assignment of `sycl::handler::MKernelName` in the
+`sycl::handler::StoreLambda` method (called from `parallel_for`) to the
+[`sycl.host.handler.set_kernel`](#syclhosthandlerset_kernel-syclsyclhosthandlersetkernel)
+operation. The rationale is that this assignment can be pattern-matched
+reliably, links the CGF to the actual kernel function, and marks a point in the
+submission process where it is guaranteed that range and argument information is
+available.
+
+The `set_kernel`'s symbol reference points to a `gpu.func` with the `kernel`
+attribute, e.g.:
+
+```mlir
+sycl.host.handler.set_kernel %handler -> @device_functions::@_ZTS10KernelName : !llvm.ptr
+```
+
+As an intermediate step to raising `set_kernel`, the SYCL dialect also defines
+the [`sycl.host.get_kernel`](#syclhostget_kernel-syclsyclhostgetkernelop)
+operation, which replaces `llvm.mlir.addressof` operations that reference a
+`gpu.func` with the `kernel` attribute.
```

Comment:
In fact, the `addressof` is to obtain the address of a string with the name of the kernel function, so we are replacing that with a reference to the actual function.

Reply:
Nice!